### PR TITLE
implement the slave_only and master_only flags

### LINKED ISSUE #203: Bug: Slave_only doesn't do anything.
the slave_only property in defaultDS probably should prevent us from becoming master...

### LINKED ISSUE #205: Bug: Master_only doesn't do anything.
We should probably implement this, at least initially, following guidance in 9.2.2.2 of the standard.

### PR DESCRIPTION
**Pull request by folkertdev**
fix #203
fix #205 

An attempt at `slave_only` and `master_only`.

### CODE REVIEW

File: statime/src/port/mod.rs [hunk 1]
```diff
@@ -451,16 +472,29 @@ impl<'a, C, F, R: Rng> Port<InBmca<'a, C, F>, R> {
...
+                        };
+
+                        self.set_forced_port_state(new_port_state);
+
                         // Immediately start sending announces and syncs
                         let duration = core::time::Duration::from_secs(0);
                         self.lifecycle.pending_action = actions![
                             PortAction::ResetAnnounceTimer { duration },
                             PortAction::ResetSyncTimer { duration }
                         ];
```
Comment (folkertdev) [inline statime/src/port/mod.rs]: it is not clear to me which actions actually need to be taken if we do a forced move into the listening state. This seems save enough to do but I don't know for sure that it is correct.

File: statime/src/port/mod.rs [hunk 2]
```diff
@@ -449,17 +470,37 @@ impl<'a, C, F, R: Rng> Port<InBmca<'a, C, F>, R> {
...
+                if default_ds.slave_only {
+                    match self.port_state {
+                        PortState::Listening => { /* do nothing */ }
+                        PortState::Slave(_) | PortState::Passive => {
+                            self.set_forced_port_state(PortState::Listening);
+
+                            // consistent with Port<InBmca>::new()
+                            let duration = self.config.announce_duration(&mut self.rng);
+                            let reset_announce = PortAction::ResetAnnounceReceiptTimer { duration };
+                            self.lifecycle.pending_action = actions![reset_announce];
```
Comment (folkertdev) [inline statime/src/port/mod.rs:482]: so this is consistent with another transition into the listening state, I think? but I can't really find what the listening state should be doing exactly. e.g. should we also reset the delay request timer here?

File: statime/src/port/mod.rs [hunk 3]
```diff
@@ -379,11 +386,15 @@ impl<'a, C, F, R: Rng> Port<InBmca<'a, C, F>, R> {
...
         current_ds: &mut CurrentDS,
         parent_ds: &mut ParentDS,
+        default_ds: &DefaultDS,
     ) -> Result<()> {
-        self.set_recommended_port_state(&recommended_state);
+        self.set_recommended_port_state(&recommended_state, default_ds);
 
         match recommended_state {
             RecommendedState::M1(defaultds) | RecommendedState::M2(defaultds) => {
+                // a slave-only PTP port should never end up in the slave state
```
Comment (davidv1992) [inline statime/src/port/mod.rs:395]: Minor typo
```suggestion
                // a slave-only PTP port should never end up in the master state
```

File: statime/src/port/mod.rs (hunk 2 above)
Comment (davidv1992) [reply statime/src/port/mod.rs:482]: This looks sensible to me. According to how I interpret the spec, listening shouldn't do much, just listen for announces.

**Review by davidv1992 — DISMISSED** [review]

**Comment by davidv1992** [comment]
Looks good, but needs a conflict resolved

**Review by davidv1992 — APPROVED** [review]

**Pull request merged**